src/layers/basic.jl

```diff
@@ -43,11 +43,18 @@ end
 
 functor(::Type{<:Chain}, c) = c.layers, ls -> Chain(ls...)
 
+function (c::Chain)(x)
+  if order() < 2
+    foldl((y,f) -> f(y), (x, c.layers...))
```

Comment:
Would it be too cheeky to do
```suggestion
    Base.afoldl(|>, x, c.layers...)
```
ref. https://github.com/JuliaLang/julia/blob/dba8f0344ab6f1a63ea4f7928610feb35f014d4f/base/operators.jl#L592.

Reply:
This was slow when I tried it :( although I did not think of `|>`

Reply:
Tidied up, here are times of some options on some very small networks:
```julia
julia> using Flux

julia> m1 = Chain(x->vcat(x,x), Dense(2,2,relu));

julia> m2 = Chain(Dense(20,30,relu), Dense(30,10), softmax);

# Tagged version

julia> @btime gradient(sum∘$m1, 3.14)
  16.584 μs (253 allocations: 9.52 KiB)
(0.33772213757038116,)

julia> @btime gradient(sum∘$m2, rand(20,20));
  42.792 μs (353 allocations: 68.00 KiB)

# foldl

julia> @eval Flux (c::Chain)(x) = foldl(|>, (x, c.layers...));

julia> Flux.Zygote.refresh()

julia> @btime gradient(sum∘$m1, 3.14)
  18.625 μs (277 allocations: 11.62 KiB)
(0.33772213757038116,)

julia> @btime gradient(sum∘$m2, rand(20,20));
  45.500 μs (385 allocations: 72.03 KiB)

# afoldl

julia> @eval Flux (c::Chain)(x) = Base.afoldl(|>, x, c.layers...);

julia> Flux.Zygote.refresh()

julia> @btime gradient(sum∘$m1, 3.14)
  84.167 μs (496 allocations: 19.03 KiB)
(0.33772213757038116,)

julia> @btime gradient(sum∘$m2, rand(20,20));
  117.917 μs (614 allocations: 77.62 KiB)

# Unstable for loop

julia> @eval Flux begin
       function (c::Chain)(x)
         y = identity(x)
         for f in c.layers
           y = f(y)
         end
         y
       end
       end

julia> Flux.Zygote.refresh()

julia> @btime gradient(sum∘$m1, 3.14)
  25.250 μs (357 allocations: 13.33 KiB)
(0.33772213757038116,)

julia> @btime gradient(sum∘$m2, rand(20,20));
  58.792 μs (512 allocations: 72.89 KiB)
  ```

Reply:
Thanks. Looks like `afoldl` would need an `rrule` like `foldl` to be remotely competitive.

Reply:
I guess it's a sad reflection on Zygote that something so simple should benefit from a complicated hand-written rule. Diffractor:
```julia
julia> fs = (sin, cos, x -> vcat(x,x), sum, exp);

julia> @btime Zygote.gradient(x -> foldl(|>, (x, $fs...)), 3)
  6.900 μs (130 allocations: 3.88 KiB)
(2.017262629514019,)

julia> @btime Zygote.gradient(x -> Base.afoldl(|>, x, $fs...), 3)
  72.083 μs (306 allocations: 11.81 KiB)
(2.017262629514019,)

julia> @btime Diffractor.gradient(x -> foldl(|>, (x, $fs...)), 3)
  9.375 μs (205 allocations: 6.52 KiB)
(2.017262629514019,)

julia> @btime Diffractor.gradient(x -> Base.afoldl(|>, x, $fs...), 3)
  3.812 μs (90 allocations: 5.56 KiB)
(2.017262629514019,)
```

Reply:
I haven't read the whole Flux documentation recently, but considering this isn't the first case of catastrophic perf degradation related to Zygote, I think it would be great if there were a simple, documented way to switch the AD Flux uses by default...

Reply:
@claforte Yea. Zygote unfortunately is just intrinsically slow for many use cases  because of its design. See [here](https://www.stochasticlifestyle.com/engineering-trade-offs-in-automatic-differentiation-from-tensorflow-and-pytorch-to-jax-and-julia/) for more details and how this is being solved.

Reply:
IIRC the major hurdle for Flux is moving away from the implicit `Params` story, which Diffractor does not plan to support. Gradients themselves do run, although it's easy to find things which break.
```
julia> m = Chain(Dense(2,3,sigmoid),sum);

julia> Zygote.gradient(m -> m([0.1, 0.2]), m)
((layers = ((weight = [0.024973915178896296 0.04994783035779259; 0.024897234881931463 0.049794469763862925; 0.02492858706809792 0.04985717413619584], bias = Float32[0.24973916, 0.24897234, 0.24928588], σ = nothing), nothing),),)

julia> Diffractor.gradient(m -> m([0.1, 0.2]), m)
(Tangent{Chain{Tuple{Dense{typeof(σ), Matrix{Float32}, Vector{Float32}}, typeof(sum)}}}(layers = (Tangent{Dense{typeof(σ), Matrix{Float32}, Vector{Float32}}}(σ = ChainRulesCore.ZeroTangent(), bias = [0.24973915178896294, 0.24897234881931463, 0.2492858706809792], weight = [0.024973915178896296 0.04994783035779259; 0.024897234881931463 0.049794469763862925; 0.02492858706809792 0.04985717413619584]), ChainRulesCore.ZeroTangent()),),)
```
 I also don't know whether benchmarks like the one above are fair to Diffractor, or if major improvements can be expected soon. Tracker seems shockingly quick there. Maybe it was a mistake for Flux to move away from that.

Reply:
https://github.com/AStupidBear/TrackerFlux.jl hasn't been updated in a few months, but could potentially be revitalized along with Tracker itself (including a ChainRules retrofit if you want to be fancy).

Support for explicit gradients is mostly tied up with [Optimisers.jl](https://github.com/FluxML/Optimisers.jl) development. The other big prereq for Diffractor support would be porting normalization layer adjoints to `rrule`s (ideally by moving the underlying functions into NNlib).

Reply:
Ok. From a look at its readme, TrackerFlux.jl stores the parameters & gradients together, as Flux used to. So the way you use it looks like the implicit parameter story. 

ReverseDiff seems to get more maintenance, and I think now uses ChainRules. Neither Tracker nor ReverseDiff seem happy with explicit non-array inputs. Although perhaps these could be overloaded to work? If so, it might be possible for Flux+Optimisers to be completely agnostic about AD, so long as `gradient(m -> loss(m,x,y), model)` returns a suitable nested structure.
```
julia> Tracker.gradient(x -> sum(x.a), (a=[1,2], b=[3,4]))
ERROR: MethodError: no method matching param(::NamedTuple{(:a, :b), Tuple{Vector{Int64}, Vector{Int64}}})

julia> ReverseDiff.gradient(x -> sum(x.a), (a=[1,2], b=[3,4]))
ERROR: MethodError: no method matching ReverseDiff.GradientConfig(::NamedTuple{(:a, :b), Tuple{Vector{Int64}, Vector{Int64}}})
```

Reply:
There's also https://github.com/dfdx/Yota.jl which seems like it could be potential pretty fast

Reply:
AD agnosticism is definitely a goal if the Optimisers.jl transition, yeah. Currently it doesn't handle Tangents, but we could steal some code from Zygote for that.

Reply:
Oh right, indeed. This does understand structs, but fails on my `foldl` tests above, not sure why
```julia
julia> Yota.grad(x -> sum(x.a), (a=[1,2], b=[3,4]))
(3, (ChainRulesCore.ZeroTangent(), Tangent{NamedTuple{(:a, :b), Tuple{Vector{Int64}, Vector{Int64}}}}(a = [1, 1],)))

julia> Yota.grad(x -> foldl(|>, (x, fs2...)), 3.0)
ERROR: No deriative rule found for op %7 = foldl(|>, %5)::Float64, try defining it using 
	ChainRulesCore.rrule(::typeof(foldl), ::typeof(|>), ::Tuple{Float64, typeof(sin), typeof(cos), var"#13#14", typeof(sum), typeof(exp)}) = ...

julia> Yota.grad(x -> Base.afoldl(|>, x, fs2...), 3.0)
ERROR: MethodError: no method matching length(::typeof(sin))
```
Cc @dfdx in case he'd like to join this hallway discussion.

Reply:
The performance is really only due to zygote unrolling chains, which isn't per se necessary. It also seems `foldl` wouldn't work generically, esp when composing layers chains of chains etc.

Reply:
@mcabbott The rule for `foldl` [exists][1] in ChainRules, so it must be a bug. I'm looking at it. Thanks for pinging me!

[1]: https://github.com/JuliaDiff/ChainRules.jl/blob/a75193768775975fac5578c89d1e5f50d7f358c2/src/rulesets/Base/mapreduce.jl#L331-L377

Reply:
Latest version of this PR replaces `foldl` with `@generated` applychain. Applied to the tiny examples above https://github.com/FluxML/Flux.jl/pull/1809#discussion_r777762381 this is faster. And it has shorter compile times (shown elsewhere).
```julia
julia> using Flux

julia> m1 = Chain(x->vcat(x,x), Dense(2,2,relu));

julia> m2 = Chain(Dense(20,30,relu), Dense(30,10), softmax);

julia> @btime gradient(sum∘$m1, 3.14)
  min 14.917 μs, mean 16.347 μs (246 allocations, 9.42 KiB)
(0.310660183429718,)

julia> @btime gradient(sum∘$m2, rand(20,20));
  min 41.417 μs, mean 50.995 μs (353 allocations, 68.27 KiB)

julia> m1 = Chain([x->vcat(x,x), Dense(2,2,relu)]);

julia> m2 = Chain([Dense(20,30,relu), Dense(30,10), softmax]);

julia> @btime gradient(sum∘$m1, 3.14)
  min 34.333 μs, mean 35.980 μs (459 allocations, 15.06 KiB)
(0.0,)

julia> @btime gradient(sum∘$m2, rand(20,20));
  min 73.500 μs, mean 85.917 μs (681 allocations, 76.66 KiB)
```
The last times here show a chain with layers stored as a vector. This is not type-stable, but helps with compile times. On very small networks like this, it is significantly slower, so it probably cannot be the default.

Reply:
FYI: latest master of Yota now correctly handles `foldl`. I haven't tested it with the latest version of this PR though since I'm currently busy speeding up the underlying tracer faster, so I'm going to wait till the PR is done and catch up later.